### ISSUE TITLE
Perform preLaunchTask in same terminal

### ISSUE THREAD

**Issue by rickstaa** [microsoft/vscode#79932]
Is it already possible to perform the prelaunch task in the same terminal as in which the debugger is started? I am creating a ROS package, and I would like to source the `devel/setup.bash` file before debugging my python node. I currently created the following `launch.json` and 'tasks.json` files:

**Launch.json**
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python Experimental: Current File (Integrated Terminal)",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "preLaunchTask": "prerun"
        },
    ]
```

**Tasks.json**
```
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "prerun",
            "type": "shell",
            "command": "source devel/setup.bash"
        }
    ]
}
```

I'm not 100% sure if this feature request belongs on this repository or the [python-extension repository](https://github.com/Microsoft/vscode-python). Feel free to correct me so I can move the issue.

**Related questions**
- [can-prelaunchtask-and-launch-start-within-the-same-terminal-in-vscode](https://stackoverflow.com/questions/57518183/can-prelaunchtask-and-launch-start-within-the-same-terminal-in-vscode)

**Comment by rickstaa** [comment]
@isidorn, @weinand & @alexr00 I was wondering if you already have an update on the status of this feature request? This feature would be very useful for me in order to source a catkin workspace before the debugger is started.

**Comment by isidorn** [comment]
This belongs in the vscode repository.
Currently it is not possible to start preLaunchTask and debug in the same terminal as far as I know. Though @weinand and @alexr00 can correct me here.
However this is a valid feature request thus adding that label and assinging to backlog.

**Comment by rickstaa** [comment]
@isidorn Thanks for the fast response and for accepting my feature request. As also stated by @sofu456 in the meantime people can achieve the desired behaviour by adding the following code to their `.bashrc`

```
# Source ros setup.bash if present
if [ -f '../devel/setup.bash' ]; then . "../devel/setup.bash";fi
```

**Comment by alexr00** [comment]
@isidorn correct, this is currently not possible.

**Comment by mickare** [comment]
Please reopen.

**Comment by rickstaa** [comment]
@mickare Maybe @isidorn can revise whether this feature-request is worth re-opening (putting it on the development back-log) or that there is not enough demand for it.

**Comment by isidorn** [comment]
I can reopen it but that does not increase the likelyhood that we do it...

**Comment by rickstaa** [comment]
@isidorn thanks for your response. By @mickare requesting it to be reopened I guess there is still support for this idea. I, therefore, think that reopening the issue till the vscodebot closes it again might be a good way to check for support.

**Comment by smalhotra3599** [comment]
Yes, please add support for this feature, it would be very useful

**Comment by madhawav** [comment]
Voting in favour of this feature. I am trying to run `vglrun bash` as a pre-launch task.

**Comment by Burznazg** [comment]
Voting for this one!

**Comment by christian-vorhemus** [comment]
+1
In my case, I set some environment variables in the preLaunchTask which should also be present in the shell that runs the debugger. So this feature would be really useful because I don't want to pollute .bashrc.

**Comment by telamonian** [comment]
@christian-vorhemus You just have to give up and use `"envFile"` to set up your env vars. It is the only way.

I've wasted far too many hours trying to find another one that works consistently

**Comment by renyhp** [comment]
Please add support for this feature, or at the very least provide a valid workaround. This is completely blocking me from debugging my program: I cannot put the script into bashrc, that would be harmful.

**Comment by weinand** [comment]
The feature asks for:
> Is it already possible to perform the prelaunch task in the same terminal as in which the debugger is started?

This is impossible for the following reason:
when the prelaunchTask is run, no debugger has started and consequently no terminal has been selected (or created).

So the only valid feature request would be:
> Run debugger in same terminal as the prelaunch task.

In this case the problem is how to communicate the terminal used by the prelaunch task into the debug adapter that actually launches the debuggee in the terminal...

**Comment by rickstaa** [comment]
@weinard That's would indeed solve  my problem. Thanks a lot for wrting it in a more clear form.

**Comment by alexr00** [comment]
@weinand feasible? Yes. However, currently the task service (which debug uses to launch the prelaunchtask) has no knowledge of terminals. I am not a huge fan of giving the task service too much knowledge of terminals since that is a nice abstraction and allows the old process task system to still work (tasks v1). It also gives tasks the freedom to someday decide that tasks shouldn't run in an integrated terminal and should instead run somewhere else.

**Comment by isidorn** [comment]
@weinand makes sense from the debug point of view if the ID of the terminal can be fetched from the `taskService`. Which @alexr00 already commented on.

**Comment by cloudhan** [comment]
I think the request should be rephrased. We want to use same terminal to do prelaunch task is that we want to use prelaunch task to do environment setup.

In short, what we need is a way to do environment setup **before debugger starts** conveniently.

So just throw away the original idea of **use prelaunch task to do the environment setup**. Give us the ability to run arbitrary code **to setup environment** is sufficient. 

If I am going to launch a bash terminal, let me source a `sh` file. 
If I am going to launch cmd, let me run a `bat` file.
If I am going to launch a powershell, let me run a `ps1` file.
And this is what should happen **in a same process**.

**Comment by cloudhan** [comment]
I think env file is sufficiently convenient. But not that flexible if you want to change it constantly.

**Comment by rickstaa** [comment]
@cloudhan I updated the title to represent the new request.

**Comment by tierriminator** [comment]
It's been a while since the last post, so I wanted to restate the importance of this issue. For me it is very common to work on projects, where a very specific setup is required before executing any code. Every time I want to use VSCode to work on those projects, one of the major pain points is setting up the debugger such that it works. For some setups, you can rely on extensions (i.e. conda), but for other setups you are left on your own. The workaround from @renyhp works astonishingly well for most cases, but must be discovered in this issue and is quite hacky.

As an example, I am currently working on a high performance cluster, which requires me to load modules into my environment before executing any programs. Additionally I need to use a custom package manager in order to load certain environment variables. I have a setup script, which does this configuration for me. Having built-in support to provide such a setup script for a debug session would have really helped me in the beginning. Instead, I just gave up on using the debugger until I found this thread with the workaround, which luckily seems to work, because the modules of the cluster are mainly loaded into environment variables.

I understand that this might be difficult to implement. But I really think that this is a common issue for many users and therefore the priority should possibly be reevaluated, considering some time has passed now.

**Comment by cadnza** [comment]
> It's been a while since the last post, so I wanted to restate the importance of this issue. For me it is very common to work on projects, where a very specific setup is required before executing any code. Every time I want to use VSCode to work on those projects, one of the major pain points is setting up the debugger such that it works. For some setups, you can rely on extensions (i.e. conda), but for other setups you are left on your own. The workaround from @renyhp works astonishingly well for most cases, but must be discovered in this issue and is quite hacky.
> 
> As an example, I am currently working on a high performance cluster, which requires me to load modules into my environment before executing any programs. Additionally I need to use a custom package manager in order to load certain environment variables. I have a setup script, which does this configuration for me. Having built-in support to provide such a setup script for a debug session would have really helped me in the beginning. Instead, I just gave up on using the debugger until I found this thread with the workaround, which luckily seems to work, because the modules of the cluster are mainly loaded into environment variables.
> 
> I understand that this might be difficult to implement. But I really think that this is a common issue for many users and therefore the priority should possibly be reevaluated, considering some time has passed now.

Hard agree with this.

**Comment by starball5** [comment]
Related on Stack Overflow: [Can preLaunchTask and launch start within the same terminal in VSCode?](https://stackoverflow.com/q/57518183/11107541). Same starting proposition of sharing terminal between preLaunchTask and debug/run, but also same backing motivation on wanting to run an environment setup script/program.

**Comment by Ryanf55** [comment]
This other Stack Overflow post is requesting a similar capability, however this comment is the same issue I observe.
* https://stackoverflow.com/questions/50458542/python-scripts-debug-in-ros-with-visual-studio-code?rq=3#comment98855212_50458793

The proposed answer does not seem to work. Here is my working version for a ROS package that imports other packages.
Because PYTHONPATH is set by running `source install/setup.bash`, that's why you need to source the local workspace.
For a ROS package where there are multiple packages, you have to change the `cwd` to the package and set the `args` to the test file because test discovery does not obey the pre-launch task either.

**tasks.json**
```json
{
   "version": "2.0.0",
   "tasks": [
      {
         "label": "gen_env",
         "type": "shell",
         "command": "source install/setup.bash && printenv > ${workspaceFolder}/.env"
      }
   ]
}
```

**launch.json**
```json
{
   "version": "0.2.0",
   "configurations": [
      {
         "name": "my_pkg tests",
         "type": "debugpy",
         "request": "launch",
         "module": "pytest",
         "args": ["-s", "test/test_mypkg.py"],
         "cwd": "${workspaceFolder}/src/my_pkg",
         "console": "integratedTerminal",
         "justMyCode": true,
         "presentation": {
          "group": "my_pkg",
         },
         "envFile": "${workspaceFolder}/.env",
         "preLaunchTask": "gen_env"
      },
}
```

As a "heavy" ROS developer with many ROS workspaces that have conflicting packages that also develops on multiple ROS distros, sourcing the ROS distro and local workspace in my `~/.bashrc` is simply not an option.
